test/context_canceled_test.go

```diff
@@ -41,38 +41,68 @@ func (s) TestContextCanceled(t *testing.T) {
 	}
 	defer ss.Stop()
 
-	var i, cntCanceled uint
-	cntPermDenied := func() uint {
-		return i - cntCanceled
-	}
-	for i, cntCanceled = 0, 0; i < 500 && (cntCanceled < 5 || cntPermDenied() < 5); i++ {
-		ctx, cancel := context.WithCancel(context.Background())
-		defer cancel()
+	// Runs 10 rounds of tests with the given delay and returns counts of status codes.
+	// Fails in case of trailer/status code inconsistency.
+	const cntRetry uint = 10
+	runTest := func(delay time.Duration) (cntCanceled, cntPermDenied uint) {
+		for i := uint(0); i < cntRetry; i++ {
+			ctx, cancel := context.WithTimeout(context.Background(), delay)
+			defer cancel()
+
+			str, err := ss.client.FullDuplexCall(ctx)
+			if err != nil {
+				continue
+			}
 
-		str, err := ss.client.FullDuplexCall(ctx)
-		if err != nil {
-			t.Fatalf("%v.FullDuplexCall(_) = _, %v, want <nil>", ss.client, err)
+			_, err = str.Recv()
+			if err == nil {
+				t.Fatalf("non-nil error expected from Recv()")
+			}
+
+			_, trlOk := str.Trailer()["a"]
+			switch status.Code(err) {
+			case codes.PermissionDenied:
+				if !trlOk {
+					t.Fatalf(`status err: %v; wanted key "a" in trailer but didn't get it`, err)
+				}
+				cntPermDenied++
+			case codes.DeadlineExceeded:
+				if trlOk {
+					t.Fatalf(`status err: %v; didn't want key "a" in trailer but got it`, err)
+				}
+				cntCanceled++
+			default:
+				t.Fatalf(`unexpected status err: %v`, err)
+			}
 		}
-		// As this duration goes up chances of Recv returning Cancelled will decrease.
-		time.Sleep(time.Duration(i) * time.Microsecond)
-		cancel()
-		_, err = str.Recv()
-		if err == nil {
-			t.Fatalf("non-nil error expected from Recv()")
+		return cntCanceled, cntPermDenied
+	}
+
+	// Tries to find the delay that causes canceled/perm denied race.
+	canceledOk, permDeniedOk := false, false
+	for lower, upper := time.Duration(0), 2*time.Millisecond; lower <= upper; {
+		delay := lower + (upper-lower)/2
+		cntCanceled, cntPermDenied := runTest(delay)
+		//log.Printf(`%v <= %v <= %v : %v, %v`, lower, delay, upper, cntCanceled, cntPermDenied)
+		if cntPermDenied > 0 && cntCanceled > 0 {
```

Comment:
These ifs could be simplified a bit:
```go
if cntPermDenied > 0 && cntCanceled > 0 {
  return
}
if cntCanceled > 0 {
  canceledOK = true
  lower += (upper-lower)/10 + 1
} else {
  permDeniedOk = true
  upper -= (upper-lower)/10 + 1
}
```

Reply:
They are not identical. If the context times out before the stream is established, then both of them (cntCanceled and cntPermDenied) can be zero. But I think I can still simplify it a bit.